relay-cogs/src/cogs.rs

```diff
@@ -0,0 +1,336 @@
+use core::fmt;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+use std::time::Instant;
+use std::usize;
+
+use crate::{utils::EnumMap, AppFeature, ResourceId};
+use crate::{CogsMeasurement, CogsRecorder, Value};
+
+/// COGS measurements collector.
+///
+/// The collector is cheap to clone.
+#[derive(Clone)]
+pub struct Cogs {
+    recorder: Arc<dyn CogsRecorder>,
+}
+
+impl Cogs {
+    /// Creates a new [`Cogs`] from a [`recorder`](CogsRecorder).
+    pub fn new<T>(recorder: T) -> Self
+    where
+        T: CogsRecorder + 'static,
+    {
+        Self {
+            recorder: Arc::new(recorder),
+        }
+    }
+
+    /// Shortcut for creating a [`Cogs`] from a [`crate::NoopRecorder`].
+    ///
+    /// All collected measurements will be dropped.
+    pub fn noop() -> Self {
+        Self::new(crate::NoopRecorder)
+    }
+}
+
+impl Cogs {
+    /// Starts a recording for a COGS measurement.
+    ///
+    /// When the returned token is dropped the measurement will be recorded
+    /// with the configured [recorder](CogsRecorder).
```

Comment:
Is it possible to generate multiple tokens at the same time? Are the features across all the tokens weighted automatically by `Cogs`, or does it double-count as if they were created sequentially?